### PR TITLE
Add NPM package

### LINKED ISSUE #3: Tracking Issue: Generate and publish `npm` package of JavaScript library.
There is no official `npm` package for this library. There are some which package it separately. It should build an artifact that be published directly to `npm`.

### PR DESCRIPTION
**Pull request by TheSpyder**
Fixes #3 

The basics should make sense but there are some possibly unexpected changes:
* I called it `@dmsnell/diff-match-patch` because https://github.com/JackuB/diff-match-patch took the top-level `diff-match-patch` name on NPM. Feel free to rename it.
* I deleted the minified code, because it's not really necessary in an NPM package. Any project that needs it will already have tools to do this.
* I added zora to run the tests, it's lightweight and fast

I'd like to also add the types from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/diff-match-patch but they're technically MIT. The types are very simple, and there's probably no legal issue with copying them, but I don't know if you want to deal with potentially mixed licensing 🤔

### CODE REVIEW

**Comment by TheSpyder** [comment]
I wasn't sure what to do with the readme in the NPM package, so I left it out. https://github.com/JackuB/diff-match-patch ships what I think is a copy of the wiki API from google.

**Comment by dmsnell** [comment]
@TheSpyder I'll be on vacation during the next week, but I'll try and get to this once I'm back. some initial thoughts:

 - are we sure this export change leaves existing code and dependencies in a good state?
 - I wonder if we can stage this in multiple pieces. one thing I thought would be nice is to start building the minified and compressed version of the library in Github Actions. it'll be nice to leave things in as they are for existing projects, but if we can remove the need to manually run build steps that would be awesome.
 - despite being slower and possibly silly, I'd probably prefer leaving major changes like revamping the tests or removing the compressed/minified version of this _out_ of a PR to start publishing an `npm` package. I get quickly overwhelmed and find it hard to focus on changes that do many things at once.

this is impressive I can tell even at a glance. great work.

**Comment by TheSpyder** [comment]
> * are we sure this export change leaves existing code and dependencies in a good state?

Hard to say. For anyone using the NPM package, it's equivalent, but I did see one person in the issues of that repo asking for a simple change to keep it running in browsers where a `module` global isn't available. I'll add that.

> * I wonder if we can stage this in multiple pieces. one thing I thought would be nice is to start building the minified and compressed version of the library in Github Actions. it'll be nice to leave things in as they are for existing projects, but if we can remove the need to manually run build steps that would be awesome.

That's definitely a good thought. I wanted to get something published for an internal deadline, but I might link the project to a branch of my fork rather than try to rush this.

> * despite being slower and possibly silly, I'd probably prefer leaving major changes like revamping the tests or removing the compressed/minified version of this _out_ of a PR to start publishing an `npm` package. I get quickly overwhelmed and find it hard to focus on changes that do many things at once.

Fair point. I'll roll back the branch to remove the test change commit and wait until the NPM package settles to submit it.

I removed the minified version mostly I figured both should be equivalent and I don't have the tools installed to generate it from the uncompressed changes I made. I can bring it back and set the NPM package to publish without it.

> this is impressive I can tell even at a glance. great work.

thank you 😁

**Comment by TheSpyder** [comment]
@dmsnell I've done a few things:
* rolled back the test change (and kept it + added types on a branch we'll use for our release). 
* made some changes to attempt to keep the browser compatibility intact.
* restored the compressed version, although it doesn't have the patches applied so the two files have diverted for the moment.

**Comment by TheSpyder** [comment]
@dmsnell will you have time to look at deploying this to NPM soon? I'm not trying to pressure you, just curious.

**Comment by dmsnell** [comment]
> will you have time to look at deploying this to NPM soon? I'm not trying to pressure you, just curious.

sorry @TheSpyder - I should have anticipated this, but I'm currently held up by some work in the current WordPress release, which is happening right now. please allot me another week and I'll try to prioritize it starting on Monday, if not over the weekend.

**Comment by TheSpyder** [comment]
Monday is fine - I think by Wednesday we might have to look at an internal deployment. We're currently using a git link to a branch of my repo (from before I rolled back the tests).

File: javascript/diff_match_patch_uncompressed.js
Comment (dmsnell) [inline javascript/diff_match_patch_uncompressed.js:78]: I believe this will change the signature of the function in a more severe way than it looks.

before the change it will return `true` for `instanceof diff_match_patch.Diff` but afterwards it won't.

probably best to preserve the existing behavior for now to avoid breaking code relying on the existing interface. it's hard to tell how far this code reaches, but it's fairly universal, so I'm sure most parts of the interface are locked.

the intended behavior can be added with `Symbol.iterator` though I'm not sure what that means for backwards compatibility either. we probably need to establish some guidelines going forward. luckily this has been supported in most browsers since 2016

File: javascript/diff_match_patch_uncompressed.js
Comment (dmsnell) [inline javascript/diff_match_patch_uncompressed.js:2406]: I could see this section being a potentially good first patch all by itself for this change. as in, for step one, could we leave everything else alone and make the module work in the existing setup _and_ make it possible for people to import it directly as an esmodule?

**Review by dmsnell — COMMENTED** [review]
@TheSpyder I'll beg your patience with me as we work through this. I'm a very slow person, and I can only fit a few things in my head at any given time.

there's a number of unrelated changes in this PR. I don't see much of a problem with most of them but I'd prefer we split them apart.

 - making the `Diff` iterable seems like a good independent change in its own right. the only concern there I have is what happens in old environments where `Symbol.iterator` isn't defined. maybe we should conditionally define the iterator function.
 - adding `index.mjs` and creating the dual export format seems awesome. we need to ensure that the existing Closure compiler process still works properly with that.

this all reminds me how important it feels to get a CI job established for building the JS module. if you want to speed things up, you can work on that before I get to it. that is, start by automating the build process of the existing repo, then once that's in place, _modify_ it so that it builds the `npm` package automatically.

or we could do this in one step, but only when ensuring that nothing breaks backwards compatbility.

maybe I'm drawing a blank, but I don't think this repository has any existing distribution channel for JavaScript other than by every project vendoring the script from source/from here.

@neilfraser did you have any official distribution for the JS code or did you expect people to download it from the Google repo?

**Comment by TheSpyder** [comment]
The code changes are all coming from https://github.com/JackuB/diff-match-patch which was adjusted to better suit use embedded in a JavaScript application, as NPM developers expect, where the Google code is based on globals and standalone use. I don't have much knowledge about them other than this setup works in our application.

The `Diff` api change, for example, was actually a rollback - a google patch in 2018 made new `Diff` tuple objects, but they're non-iterable which breaks ES6 environments. So it was swapped back to a plain array.
https://github.com/JackuB/diff-match-patch/issues/14

> start by automating the build process of the existing repo

I admit I have no idea what that is 😂  I have only looked seriously at the javascript folder.

File: javascript/diff_match_patch_uncompressed.js
Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js:2406]: I can pull this out to a new PR if you like. It will help with CommonJS deployments, but the diff API will still have issues in an ES6 environment as I mentioned in my other comment.

**Comment by dmsnell** [comment]
> this all reminds me how important it feels to get a CI job established for building the JS module. if you want to speed things up, you can work on that before I get to it. that is, start by automating the build process of the existing repo, then once that's in place, _modify_ it so that it builds the `npm` package automatically.
> 
> or we could do this in one step, but only when ensuring that nothing breaks backwards compatbility.

on second thought this change of repo is a great opportunity to break from all the constraints from `diff-match-patch`, like Closure compiler. I forget that _nobody_ has this official repo in their `npm` projects because it's never published one.

let's see if we can get a CI job setup which produces a minified build, and let's compare that output to what Closure Compiler has been producing. I'm inclined to try something like biomejs which doesn't pull in a bunch of dependencies.

File: javascript/diff_match_patch_uncompressed.js
Comment (dmsnell) [inline javascript/diff_match_patch_uncompressed.js:2390]: who is @ForbesLindesay and why did they add this code?
I think this would be better suited as a `git` commit or a `Co-authored-by:` trailer on the commit, given that it's trying to attribute the change but the comment doesn't indicate why it's there

File: javascript/package.json
Comment (dmsnell) [inline javascript/package.json:4]: We shouldn't need to refer to the source of the fork here. Effectively this is a new project now. 🤷‍♂️ 

maybe something like "Efficient and human-readable text diffs with helpers."?

File: javascript/diff_match_patch_uncompressed.js
Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js:78]: Symbol.iterator was mentioned in the original issue about this, it's been a part of every JavaScript environment except IE11 for probably 7 years. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator#browser_compatibility

Microsoft don't support IE11 anymore, I suspect anyone still deploying for it is already transpiling their code 🤷‍♂️

File: javascript/diff_match_patch_uncompressed.js
Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js:2390]: The original developer of the repository these changes came from, scroll down to 2015
https://github.com/JackuB/diff-match-patch/commits/master/

**Comment by TheSpyder** [comment]
> I forget that _nobody_ has this official repo in their `npm` projects because it's never published one.

That's the key I'm focusing on - anyone using this library on NPM is doing `npm install diff-match-patch` and perhaps assuming it's the same, as we did, but it's not.
https://www.npmjs.com/package/diff-match-patch

File: javascript/diff_match_patch_uncompressed.js
Comment (dmsnell) [reply javascript/diff_match_patch_uncompressed.js:78]: yeah let's leave it in then. if anyone is importing this into `npm` projects they should have something supported within the past eight years.

later on I'd be for converting the `Function` interfaces to `class` syntax. for now though we can keep the `Diff` identity while offering destructuring

File: javascript/diff_match_patch_uncompressed.js
Comment (dmsnell) [reply javascript/diff_match_patch_uncompressed.js:2406]: could we leave it as ESModule syntax in the code and have our build tool generate a CommonJS module?

if we _break_ with the vendored file of the past I don't see why we'd need to make Closure compiler happy

File: .gitignore
Comment (dmsnell) [inline .gitignore:1]: let's leave this out of here since the project isn't specific to JetBrains software. you can add a global `.gitignore` in your home directory to avoid including files that are incidentally to the tools you yourself use

File: javascript/diff_match_patch_uncompressed.js
Comment (dmsnell) [inline javascript/diff_match_patch_uncompressed.js:89]: as with the previous comment, we should leave these lines in so we don't break the identify of the `Diff` object.

File: javascript/diff_match_patch_uncompressed.js
Comment (dmsnell) [reply javascript/diff_match_patch_uncompressed.js:2390]: What if we eliminated all of this and instead used only module syntax?

```js
export default diff_match_patch;
export diff_match_path;
export DIFF_DELETE;
export DIFF_INSERT;
export DIFF_EQUAL;
```

then we omit `index.mjs` and instead, generate a Common-JS artifact? or not. maybe it's too late, but I don't even know if I care about Common-JS, if we're breaking free in this repo. won't any project with a recent build _have_ to support ESM? I was thinking that several major libraries are now exporting their packages untranspiled

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js:2390]: There are plenty of CJS projects around, it's still the NodeJS default, and using `module.exports` here alongside the `mjs` file (using with the references in `package.json` to differentiate them) means it can work for both.

I'm happy to make this ESM-only, if that's what you want - I certainly only need ESM.

File: .gitignore
Comment (TheSpyder) [reply .gitignore:1]: This is a habit, sorry, I only recently switched to webstorm and I _do_ check the files in for some of my other projects so I don't use a global ignore. Happy to remove.

File: javascript/diff_match_patch_uncompressed.js
Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js:89]: OK, so to be clear you want to keep the Diff object and make it iterable?

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js:89]: I'm circling back to this PR and realised I shelved the work until this question was answered and then forgot to follow up. Is my statement above correct?

Comment (dmsnell) [reply javascript/diff_match_patch_uncompressed.js:89]: Yeah sorry for not responding earlier. Since this is a change about build-time semantics I think we should avoid changing runtime semantics. That is, if someone was already using these functions we don't want to inadvertently break their code just because we were trying to make it easier to import the library.

Comment (dmsnell) [reply javascript/diff_match_patch_uncompressed.js:89]: was it previously iterable? I think what happened before is that we were setting the properties `0` and `1` on the object and it was possible to read those. it's array-like, but only because those properties are defined. I think it would still failure in an iteration like `diff.forEach()` (maybe I'm wrong, should be easy to confirm).

the biggest change between `return [a, b]` and `this[0] = a; this[1] = b` is that we're taking something that _was_ a class (or more specifically, an object with a specific constructor `Diff`), and turning it into a semantically void array of two elements.

we can preserve this while making it iterable to destructure it. maybe it would all look clearer using the `class` syntax.

```js
class Diff {
	constructor( op, text ) {
		this[0] = op;
		this[1] = text;
	}

	*[Symbol.iterator]() {
		yield this[0];
		yield this[1];
	}
}
```

(this example may be entirely broken)

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js:89]: Yes, the specific reason this needs to be iterable is for destructuring, e.g. 
```
var a = dmp.diff_main('abc', 'ab123c', false)
const [eq, str] = a[0]
```
Which doesn't work now because `a[0]` is a `Diff` object. 

Whether it's a breaking change depends on who you think the target audience is. NPM users are IMO likely using the code I'm submitting here which came from the existing (now-abandoned) NPM package I mentioned above.
https://www.npmjs.com/package/diff-match-patch

As far as I can see, the history of the code is:
* Google used an array for a long time
* In 2018 they [swapped it to this `Diff` object](https://github.com/google/diff-match-patch/commit/cd60d246aecf871367313c7cf6dc8814af32c5e3), which isn't iterable
* The developer of the now-abandoned NPM package logged an issue about it as it's a breaking change https://github.com/google/diff-match-patch/issues/39
* In the meantime, that developer [reversed the change and made it an array again for NPM](https://github.com/JackuB/diff-match-patch/commit/18711ba8abef83aba02d30fe20206fad6d5c5487) - at the time I think this was expected to be temporary.

It looks like your suggested fix matches the one mentioned in that ticket on the google repo:
https://github.com/google/diff-match-patch/issues/39#issuecomment-457262957

So let's just include that.

File: javascript/diff_match_patch_uncompressed.js
Comment (dmsnell) [inline javascript/diff_match_patch_uncompressed.js:72]: we can remove this change now, or remove the old code as well. it should be evident to folks that `Diff` is its own class, and the use of `Symbol.iterator` should be an even more clear indicator that this is array-like.

**Comment by dmsnell** [comment]
@TheSpyder In #9 I eagerly added the `Symbol.iterator` method so hopefully that makes this PR a bit simpler. You are welcome to incorporate that change by merge or rebase or however you want.

File: javascript/diff_match_patch_uncompressed.js
Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js:2432]: From what I've seen, build tools seem to do just something like the `index.mjs` file in this PR. Projects that support both do it as I've done in this PR, although I certainly won't complain if you want to make it ESM-only.

**Comment by TheSpyder** [comment]
@dmsnell I think I've covered most of the feedback now - it's just the ESM question (whether to make it the default, whether to make it a new PR) that I'm not sure about.

**Comment by dmsnell** [comment]
@TheSpyder I'll try and get back to this in the next few days after thinking more about ESM. I'm tempted to be a bit aggressive and focus more on the future direction and letting new code accept this. Seems like most projects today can and should be able to handle an ESM package. If it's not clear, we can merge and iterate later. Thanks for your patience with me again. It's been a very busy couple months on my end and I didn't mean to let this go so long.

**Comment by TheSpyder** [comment]
@dmsnell ok, what I might do in the meantime is update the testing code I pulled out of this PR and make a new PR with it. There's not really a rush on publishing this, although my team has been asking me why we're still deploying production code that depends on my personal github fork of this project 😅

File: javascript/package.json
Comment (sirreal) [inline javascript/package.json:12]: If we were to go with both forms, it looks like this is flipped - [import should be esm and require should be cjs](https://nodejs.org/api/packages.html#conditional-exports):

```suggestion
    "import": "./index.mjs",
    "require": "./diff_match_patch_uncompressed.js"
```

File: javascript/package.json
Comment (sirreal) [inline javascript/package.json:9]: [Let's declare it a module](https://nodejs.org/api/packages.html#type):

```suggestion
  ],
  "type": "module",
```

**Review by sirreal — COMMENTED** [review]
If I were to publish this today as a brand new package, I think I'd stick with ESM-only and skip all the CJS work. That seems like eagerly taking on technical debt when there's an opportunity to avoid it.

If sufficient demand for a CJS version were to manifest, that could always be explored later.

File: javascript/package.json
Comment (TheSpyder) [reply javascript/package.json:12]: Whoops! I wonder why my tooling didn't pick this up. I guess bundlers are just used to things being weird and make it work somehow.

File: javascript/package.json
Comment (TheSpyder) [reply javascript/package.json:9]: I'm working on a PR to add types, maybe that would be a good time to make the switch to ESM.

**Comment by TheSpyder** [comment]
> It's plausible that all this can all still work with a swap to `new Diff()`, but I'm not entirely confident about that change.

It wasn't hard to do and seems to work, so I'm happy to go with the object.

**Comment by TheSpyder** [comment]
@dmsnell in an earlier comment you mentioned perhaps the ESM change should be a separate PR. I'm happy to do that, if you decide that's the path forward, or extract it to a new PR.

I've created new branches to add types and tests, as separate PRs, once this change is approved and merged. They're working very well on my team's project.

File: javascript/diff_match_patch_uncompressed.js
Comment (dmsnell) [inline javascript/diff_match_patch_uncompressed.js:2422]: @TheSpyder you've been so persistent and so helpful on this. I wish I had more ability to quickly respond.

in reflection I think it's best for us to lean hard into ES Modules. let's eliminate all of the CommonJS references and only create ESM output.

I've been wanting to do this myself instead of asking you to make the change, but that has only resulted in me deferring this work. so, if you get this updated we'll get it merged, otherwise you can still wait on me and I'll merge it when I can.

Comment (dmsnell) [reply javascript/diff_match_patch_uncompressed.js]: @TheSpyder I've pushed a commit and may have messed up everything. Would you be interested in confirming if my changes work for your project?

I haven't fixed the tests, and I don't like that, but I'm willing to let that remain a follow-up issue because I want to overhaul the tests anyway; take them out of the browser and into a more raw test runner.

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js]: That's fine! I understand. I'll take a look soon, Monday at the latest.

I have already converted the JS tests to a nodejs runner, that's what I pulled out of this PR at your request. Once it's cleaned up and merged that will go into PR as well 🙂

Comment (dmsnell) [reply javascript/diff_match_patch_uncompressed.js]: yeah sorry. I'm horribly inconsistent and slow at times.

if you want, throw out what I just pushed - which is the start of rewriting the tests. I'd almost prefer we do all the tests in a different change, but I did want to make sure it's viable to make the switch, so tossed out Bun and a standard runner/spec file.

let me know your thoughts.

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js]: I used zora which is extremely lightweight. I know Jest is popular, but I find it too heavy for my liking :)

My commit is here:
https://github.com/TheSpyder/diff-match-patch/commit/3d8b914b05d966e2f31a5810b468ae7e564da237

It focuses on minimal changes in the diff, to preserve the tests themselves and only need reviewing on the test runner.

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js]: > yeah sorry. I'm horribly inconsistent and slow at times.

I will never judge you for taking your time, or changing your mind. I fully understand the pressures of maintaining something open source like this 🙂

Comment (sirreal) [reply javascript/diff_match_patch_uncompressed.js]: For tests, [have you considered the node test runner?](https://nodejs.org/docs/latest-v20.x/api/test.html#test-runner)

Comment (dmsnell) [reply javascript/diff_match_patch_uncompressed.js]: Thanks @sirreal, I forgot that `node` itself has a runner. I think that would be best and I chose `bun` specifically because it didn't require installing a separate testing library.

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js]: I'll take a look. A library can make writing tests easier, but I don't know that we will be adding new tests very often.

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js]: I'm going to revert your changes @dmsnell and redo them - I'd like to make sure I can prove the diff made as few edits as possible, which means avoiding code formatting. We can apply that later. This will also force push away the bun/jest code while I investigate swapping to the node test runner.

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js]: I saved a local copy of the branch in case we need it.

Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js]: So of course after all that I have changed to webstorm since opening this PR and it reformatted some code 🤦 sorry about the extra force pushes.

**Comment by TheSpyder** [comment]
@dmsnell now that we've brought the test changes back in, can we bring the types as well? Or should that still be a separate PR?
https://github.com/dmsnell/diff-match-patch/commit/7e93e49d258c354c09f63087e06dde6d2a3428de

File: javascript/.gitignore
Comment (dmsnell) [inline javascript/.gitignore:3]: this doesn't belong here

**Review by dmsnell — COMMENTED** [review]
Nice. I like what you did with the tests, and actually I think the whitespace is fine. I'll merge this after another final review and we can move forward with a new base.

Awesome work!

File: javascript/diff_match_patch_uncompressed.js
Comment (TheSpyder) [reply javascript/diff_match_patch_uncompressed.js]: I know, but not everyone does 😂

File: javascript/.gitignore
Comment (TheSpyder) [reply javascript/.gitignore:3]: It's there because `npm pack` which I use to verify the package contents makes a file matching that pattern, and I always forget to delete it before doing `git add .`

feel free to remove it before merging.

File: javascript/index.mjs
Comment (sirreal) [inline javascript/index.mjs:26]: The package is now declared `"type":"module"` so it should not be necessary to rename this file to `.mjs` extension now.

**Review by sirreal — APPROVED** [review]
This looks good in general, well done, thanks for persevering.

I've left some minor notes but nothing big.

One interesting thing I noticed when exploring the tests is that `Diff` objects behave like arrays in some ways, but are not arrays. There are a lot of test assertions that abuse this and compare tuples with diff objects, e.g. `[ DIFF_EQUAL, "" ] === new Diff( DIFF_EQUAL, "" )`. If we replace the array shorthand with `Diff` object creation, then we can replace the recursive equality check function with `deepEqual`.

File: javascript/tests/diff_match_patch_test.js
Comment (TheSpyder) [reply javascript/tests/diff_match_patch_test.js:2]: ah yeah this was left over from before it switched to module, thanks

File: javascript/tests/diff_match_patch_test.js
Comment (TheSpyder) [reply javascript/tests/diff_match_patch_test.js:3]: oh interesting 🤔 thanks for the tip!

File: javascript/index.mjs
Comment (TheSpyder) [reply javascript/index.mjs:26]: not technically, but I figured it didn't hurt 🤔  happy to swap it back though

**Comment by TheSpyder** [comment]
> One interesting thing I noticed when exploring the tests is that `Diff` objects behave like arrays in some ways, but are not arrays. There are a lot of test assertions that abuse this and compare tuples with diff objects, e.g. `[ DIFF_EQUAL, "" ] === new Diff( DIFF_EQUAL, "" )`. If we replace the array shorthand with `Diff` object creation, then we can replace the recursive equality check function with `deepEqual`.

It's a result of a change that was made years ago; the diff used to be an array, google changed it to an object. The tests haven't changed since then. For this PR, I want the absolute minimal changes to get it on NPM - perhaps log that as a new issue?

**Comment by TheSpyder** [comment]
@dmsnell I think I have a github action that will run the node tests, but probably best to save that for a new PR 🤔

**Comment by dmsnell** [comment]
Yeah @TheSpyder let's suggest those CI runs in a separate PR. Bravo! Thank you for staying persistent with this.

https://www.npmjs.com/package/@dmsnell/diff-match-patch

Feel free to start using it, and let's continue making it better!

**Comment by TheSpyder** [comment]
Fantastic! thank you

It will be easier for my team to use with types, I might make two PRs 🫣

**Pull request merged**
